File: backend/app/services/expenses.py
"""Expense claims: submit with receipts, approved by the team lead and then by HR. No HTTP here.

    submit ──▶ pending_lead ──lead approves──▶ pending_hr ──HR approves──▶ approved
    (no lead ─────────────────────────────────▲)
               either pending step ──reject (with a reason)──▶ rejected
               either pending step ──the submitter withdraws──▶ withdrawn

Who decides:
- pending_lead: the submitter's team lead
- pending_hr: HR users, and admins (the privileged accounts). They're told by notification;
  if nobody has the HR role yet, the admins are.
- Nobody decides their own expense, and the HR step needs a different person than the lead
  who approved it (two different people approve every expense).
"""

from dataclasses import dataclass, field
from datetime import UTC, datetime, timedelta

from sqlalchemy import or_, select
from sqlalchemy.orm import Session, joinedload

from app import files
from app.config import settings
from app.email import Email
from app.errors import Conflict, Forbidden, NotFound, ValidationFailed
from app.models import Expense, ExpenseReceipt, ExpenseStatus, NotificationType, User
from app.schemas.expense import ExpenseCreate
from app.services import notifications
from app.services.seats import office_today

S = ExpenseStatus

MAX_RECEIPTS = 5
MAX_RECEIPT_BYTES = 5 * 1024 * 1024
OLDEST_DAYS = 90  # a receipt older than this is too late to claim

RECEIPT_TYPES = {
    ".pdf": ("application/pdf", files.is_pdf),
    ".png": ("image/png", files.is_png),
    ".jpg": ("image/jpeg", files.is_jpeg),
    ".jpeg": ("image/jpeg", files.is_jpeg),
    ".webp": ("image/webp", files.is_webp),
}

ALLOWED_MOVES: dict[ExpenseStatus, set[ExpenseStatus]] = {
    S.PENDING_LEAD: {S.PENDING_HR, S.REJECTED, S.WITHDRAWN},
    S.PENDING_HR: {S.APPROVED, S.REJECTED, S.WITHDRAWN},
    S.APPROVED: set(),
    S.REJECTED: set(),
    S.WITHDRAWN: set(),
}


def check_move(current: ExpenseStatus, target: ExpenseStatus) -> None:
    if target not in ALLOWED_MOVES[current]:
        code = "not_withdrawable" if target == S.WITHDRAWN else "not_pending"
        raise Conflict(code, f"This expense is already {current.replace('_', ' ')}")


def money(expense: Expense) -> str:
    return f"€{expense.amount:,.2f}"


# --- who ---


def is_hr_decider(user: User) -> bool:
    return user.is_hr or user.is_admin


def hr_recipients(db: Session, exclude: set[int]) -> list[User]:
    """Who hears about an expense waiting for HR: the HR users, or the admins while there are none."""
    hr = [u for u in db.scalars(select(User).where(User.is_hr)) if u.id not in exclude]
    if hr:
        return hr
    return [u for u in db.scalars(select(User).where(User.is_admin)) if u.id not in exclude]


def can_see(user: User, expense: Expense) -> bool:
    owner = expense.user
    return (
        owner.id == user.id
        or owner.team_lead_id == user.id
        or expense.lead_decided_by_id == user.id
        or is_hr_decider(user)
    )


def waiting_for(expense: Expense) -> str | None:
    if expense.status == S.PENDING_LEAD:
        return expense.user.team_lead.name if expense.user.team_lead else "Team lead"
    if expense.status == S.PENDING_HR:
        return "HR"
    return None


# --- submitting ---


@dataclass
class Outcome:
    """What a service call did, plus the emails the router sends after the response."""

    expense: Expense
    emails: list[Email] = field(default_factory=list)


def _receipt(filename: str | None, data: bytes) -> ExpenseReceipt:
    name = files.clean_filename(filename)
    kind = RECEIPT_TYPES.get(files.extension(name))
    if kind is None:
        raise ValidationFailed("receipts", "Receipts must be PDF, PNG, JPEG or WebP files", "receipt_type", name)
    if not data:
        raise ValidationFailed("receipts", f"{name} is empty", "receipt_empty", name)
    if len(data) > MAX_RECEIPT_BYTES:
        raise ValidationFailed("receipts", f"{name} is too large (max 5 MB)", "receipt_too_large", name)
    content_type, looks_right = kind
    if not looks_right(data):
        raise ValidationFailed("receipts", f"{name} doesn't look like a {files.extension(name)} file", "receipt_mismatch", name)
    return ExpenseReceipt(filename=name, content_type=content_type, size=len(data), data=data)


def _emails_to(deciders: list[User], expense: Expense, why: str) -> list[Email]:
    return [
        Email(
            to=decider.email,
            subject=f"Expense to approve: {expense.user.name}, {money(expense)}",
            body=(
                f"Hi {decider.name},\n\n{why}\n\n"
                f"Approve or reject it here: {settings.app_url}/approvals\n\n— PreyingMantis"
            ),
        )
        for decider in deciders
    ]


def submit(db: Session, user: User, data: ExpenseCreate, uploads: list[tuple[str | None, bytes]]) -> Outcome:
    """`uploads` is (filename, bytes) per receipt; each may be one byte over the limit, to detect it."""
    today = office_today()
    if data.spent_on > today:
        raise ValidationFailed("spent_on", "The date can't be in the future", "spent_in_future", str(data.spent_on))
    if data.spent_on < today - timedelta(days=OLDEST_DAYS):
        raise ValidationFailed(
            "spent_on", f"Expenses older than {OLDEST_DAYS} days can't be claimed", "spent_too_long_ago", str(data.spent_on)
        )
    if not 1 <= len(uploads) <= MAX_RECEIPTS:
        raise ValidationFailed("receipts", f"Add 1 to {MAX_RECEIPTS} receipts", "receipt_count", len(uploads))

    expense = Expense(
        user=user,
        title=data.title,
        description=data.description or None,
        category=data.category,
        amount=data.amount,
        spent_on=data.spent_on,
        status=S.PENDING_LEAD if user.team_lead is not None else S.PENDING_HR,
        receipts=[_receipt(name, content) for name, content in uploads],
    )
    db.add(expense)
    if expense.status == S.PENDING_LEAD:
        deciders = [user.team_lead]
        message = f"{user.name} submitted an expense: {expense.title} ({money(expense)})"
    else:
        deciders = hr_recipients(db, exclude={user.id})
        message = f"{user.name} submitted an expense for HR approval: {expense.title} ({money(expense)})"
    notifications.notify(db, deciders, NotificationType.EXPENSE_SUBMITTED, message, link="/approvals")
    db.commit()  # the expense, its receipts and the notifications together
    db.refresh(expense)
    return Outcome(expense, _emails_to(deciders, expense, message + "."))


# --- reading ---


def _with_people(query):
    # Everything ExpenseRead shows, in the same query: the submitter (+ their lead and avatar) and the deciders
    return query.options(
        joinedload(Expense.user).joinedload(User.team_lead),
        joinedload(Expense.user).joinedload(User.avatar),
        joinedload(Expense.lead_decided_by).joinedload(User.avatar),
        joinedload(Expense.hr_decided_by).joinedload(User.avatar),
    )


def mine(db: Session, user: User) -> list[Expense]:
    """My expenses, newest first."""
    query = select(Expense).where(Expense.user_id == user.id).order_by(Expense.submitted_at.desc(), Expense.id.desc())
    return list(db.scalars(_with_people(query)).unique())


def get(db: Session, user: User, expense_id: int) -> Expense:
    """The expense, or 404 when it doesn't exist or the user may not see it."""
    expense = db.scalars(_with_people(select(Expense).where(Expense.id == expense_id))).unique().one_or_none()
    if expense is None or not can_see(user, expense):
        raise NotFound("Expense not found")
    return expense


def to_decide(db: Session, user: User) -> list[Expense]:
    """What waits for this user: their reports' expenses at the lead step, and (for HR and admins)
    everyone else's at the HR step, except ones they approved as lead. Oldest first."""
    Owner = User
    for_me = (Expense.status == S.PENDING_LEAD) & (Owner.team_lead_id == user.id)
    if is_hr_decider(user):
        hr_step = (
            (Expense.status == S.PENDING_HR)
            & (Expense.user_id != user.id)
            & or_(Expense.lead_decided_by_id.is_(None), Expense.lead_decided_by_id != user.id)
        )
        for_me = or_(for_me, hr_step)
    query = (
        select(Expense)
        .join(Owner, Expense.user_id == Owner.id)
        .where(for_me)
        .order_by(Expense.submitted_at, Expense.id)
    )
    return list(db.scalars(_with_people(query)).unique())


def receipt(db: Session, user: User, expense_id: int, receipt_id: int) -> ExpenseReceipt:
    expense = get(db, user, expense_id)
    found = next((r for r in expense.receipts if r.id == receipt_id), None)
    if found is None:
        raise NotFound("Receipt not found")
    return found


# --- deciding ---


def _lock(db: Session, user: User, expense_id: int) -> Expense:
    """Checks visibility, then locks the row and re-reads it, so two decisions at once are one after the other."""
    get(db, user, expense_id)
    return db.get(Expense, expense_id, with_for_update=True, populate_existing=True)


def _check_decider(user: User, expense: Expense) -> None:
    if expense.user_id == user.id:
        raise Forbidden("You can't decide your own expense")
    if expense.status == S.PENDING_LEAD and expense.user.team_lead_id != user.id:
        raise Forbidden("Only the submitter's team lead can decide this step")
    if expense.status == S.PENDING_HR:
        if not is_hr_decider(user):
            raise Forbidden("Only HR can decide this step")
        if expense.lead_decided_by_id == user.id:
            raise Conflict("second_approver_needed", "You approved this as team lead: someone else in HR gives the second approval")


def approve(db: Session, user: User, expense_id: int) -> Outcome:
    expense = _lock(db, user, expense_id)
    if expense.status not in (S.PENDING_LEAD, S.PENDING_HR):
        check_move(expense.status, S.APPROVED)  # raises not_pending
    _check_decider(user, expense)
    now = datetime.now(UTC)
    emails: list[Email] = []
    if expense.status == S.PENDING_LEAD:
        check_move(expense.status, S.PENDING_HR)
        expense.status = S.PENDING_HR
        expense.lead_decided_by, expense.lead_decided_at = user, now
        notifications.notify(
            db,
            [expense.user],
            NotificationType.EXPENSE_LEAD_APPROVED,
            f"{user.name} approved your expense {expense.title}: it's with HR now",
            link=f"/expenses/{expense.id}",
        )
        hr = hr_recipients(db, exclude={expense.user_id, user.id})
        message = f"{expense.user.name}'s expense {expense.title} ({money(expense)}) was approved by {user.name} and needs HR approval"
        notifications.notify(db, hr, NotificationType.EXPENSE_SUBMITTED, message, link="/approvals")
        emails = _emails_to(hr, expense, message + ".")
    else:
        check_move(expense.status, S.APPROVED)
        expense.status = S.APPROVED
        expense.hr_decided_by, expense.hr_decided_at = user, now
        notifications.notify(
            db,
            [expense.user],
            NotificationType.EXPENSE_APPROVED,
            f"Your expense {expense.title} ({money(expense)}) was approved",
            link=f"/expenses/{expense.id}",
        )
    db.commit()  # the decision and its notifications together
    db.refresh(expense)
    return Outcome(expense, emails)


def reject(db: Session, user: User, expense_id: int, reason: str) -> Expense:
    expense = _lock(db, user, expense_id)
    check_move(expense.status, S.REJECTED)
    _check_decider(user, expense)
    now = datetime.now(UTC)
    if expense.status == S.PENDING_LEAD:
        expense.lead_decided_by, expense.lead_decided_at = user, now
    else:
        expense.hr_decided_by, expense.hr_decided_at = user, now
    expense.status = S.REJECTED
    expense.rejection_reason = reason
    notifications.notify(
        db,
        [expense.user],
        NotificationType.EXPENSE_REJECTED,
        f"{user.name} rejected your expense {expense.title}: {reason}",
        link=f"/expenses/{expense.id}",
    )
    db.commit()
    db.refresh(expense)
    return expense


def withdraw(db: Session, user: User, expense_id: int) -> Expense:
    expense = _lock(db, user, expense_id)
    if expense.user_id != user.id:
        raise Forbidden("You can only withdraw your own expenses")
    check_move(expense.status, S.WITHDRAWN)
    expense.status = S.WITHDRAWN
    db.commit()
    db.refresh(expense)
    return expense

File: frontend/src/mocks/data/expenses.ts
import type { Expense, ExpenseCategory, ExpenseStatus } from '../../api/expenses'
import { mockAvatarUrl } from './avatars'
import { findSeedUserByEmail, findSeedUserById } from './users'

// Same rules as backend/app/services/expenses.py: team lead first, then HR (HR users and admins).
type Viewer = { id: number; is_admin: boolean }
type StoredReceipt = { id: number; filename: string; content_type: string; size: number; data: Uint8Array<ArrayBuffer> }
type Stored = Omit<Expense, 'user' | 'waiting_for' | 'lead_decision' | 'hr_decision' | 'receipts'> & {
  userId: number
  lead: { byId: number; at: string } | null
  hr: { byId: number; at: string } | null
  receipts: StoredReceipt[]
}

const store: Stored[] = []
let nextId = 1
let nextReceiptId = 1

const person = (id: number) => {
  const user = findSeedUserById(id)
  return { id, name: user?.name ?? 'Unknown', avatar_url: mockAvatarUrl(id) }
}
const leadOf = (userId: number) => {
  const email = findSeedUserById(userId)?.teamLeadEmail
  return email ? findSeedUserByEmail(email) : undefined
}
const isHrDecider = (viewer: Viewer) => viewer.is_admin || Boolean(findSeedUserById(viewer.id)?.is_hr)

function toExpense(s: Stored): Expense {
  const { userId, lead, hr, receipts, ...fields } = s
  const rejectedByLead = s.status === 'rejected' && !hr
  return {
    ...fields,
    user: person(userId),
    waiting_for: s.status === 'pending_lead' ? (leadOf(userId)?.name ?? 'Team lead') : s.status === 'pending_hr' ? 'HR' : null,
    lead_decision: lead ? { by: person(lead.byId), at: lead.at, approved: !rejectedByLead } : null,
    hr_decision: hr ? { by: person(hr.byId), at: hr.at, approved: s.status === 'approved' } : null,
    receipts: receipts.map(({ data: _data, ...r }) => r),
  }
}

const canSee = (viewer: Viewer, s: Stored) =>
  s.userId === viewer.id || leadOf(s.userId)?.id === viewer.id || s.lead?.byId === viewer.id || isHrDecider(viewer)

export const listMockMyExpenses = (viewerId: number) =>
  store.filter((s) => s.userId === viewerId).sort((a, b) => b.submitted_at.localeCompare(a.submitted_at)).map(toExpense)

export function getMockExpense(viewer: Viewer, id: number) {
  const found = store.find((s) => s.id === id)
  return found && canSee(viewer, found) ? found : undefined
}

export const mockExpenseRead = toExpense

export function listMockExpenseApprovals(viewer: Viewer) {
  return store
    .filter(
      (s) =>
        (s.status === 'pending_lead' && leadOf(s.userId)?.id === viewer.id) ||
        (s.status === 'pending_hr' && isHrDecider(viewer) && s.userId !== viewer.id && s.lead?.byId !== viewer.id),
    )
    .sort((a, b) => a.submitted_at.localeCompare(b.submitted_at))
    .map(toExpense)
}

const RECEIPT_TYPES: Record<string, string> = {
  pdf: 'application/pdf',
  png: 'image/png',
  jpg: 'image/jpeg',
  jpeg: 'image/jpeg',
  webp: 'image/webp',
}

type Fields = { title: string; description: string; category: ExpenseCategory; amount: string; spent_on: string }
type Upload = { filename: string; data: Uint8Array<ArrayBuffer> }

function invalid(field: string, type: string, msg: string) {
  return { status: 422 as const, detail: [{ type, loc: ['body', field], msg }] }
}

// POST /api/expenses (the backend's main checks; the file bytes aren't sniffed here)
export function submitMockExpense(userId: number, fields: Fields, uploads: Upload[]) {
  if (!fields.title?.trim()) return invalid('title', 'string_too_short', 'String should have at least 1 character')
  if (!/^\d{1,5}(\.\d{1,2})?$/.test(fields.amount) || Number(fields.amount) <= 0) {
    return invalid('amount', 'decimal_parsing', 'Enter an amount like 12.50')
  }
  if (uploads.length < 1 || uploads.length > 5) return invalid('receipts', 'receipt_count', 'Add 1 to 5 receipts')
  const receipts: StoredReceipt[] = []
  for (const upload of uploads) {
    const type = RECEIPT_TYPES[upload.filename.split('.').pop()?.toLowerCase() ?? '']
    if (!type) return invalid('receipts', 'receipt_type', 'Receipts must be PDF, PNG, JPEG or WebP files')
    receipts.push({ id: nextReceiptId++, filename: upload.filename, content_type: type, size: upload.data.length, data: upload.data })
  }
  const stored: Stored = {
    id: nextId++,
    userId,
    title: fields.title.trim(),
    description: fields.description?.trim() || null,
    category: fields.category,
    amount: Number(fields.amount).toFixed(2),
    currency: 'EUR',
    spent_on: fields.spent_on,
    status: leadOf(userId) ? 'pending_lead' : 'pending_hr',
    submitted_at: new Date().toISOString(),
    lead: null,
    hr: null,
    rejection_reason: null,
    receipts,
  }
  store.push(stored)
  return { status: 201 as const, expense: toExpense(stored) }
}

// POST …/approve | reject | withdraw
export function decideMockExpense(viewer: Viewer, id: number, action: 'approve' | 'reject' | 'withdraw', reason?: string) {
  const s = getMockExpense(viewer, id)
  if (!s) return { status: 404 as const, detail: 'Expense not found' }
  const refuse = (code: string, message: string) => ({ status: 409 as const, detail: { code, message } })
  const forbid = (detail: string) => ({ status: 403 as const, detail })
  const pending = s.status === 'pending_lead' || s.status === 'pending_hr'
  const now = new Date().toISOString()

  if (action === 'withdraw') {
    if (s.userId !== viewer.id) return forbid('You can only withdraw your own expenses')
    if (!pending) return refuse('not_withdrawable', `This expense is already ${s.status}`)
    s.status = 'withdrawn'
    return { status: 200 as const, expense: toExpense(s) }
  }
  if (!pending) return refuse('not_pending', `This expense is already ${s.status}`)
  if (s.userId === viewer.id) return forbid("You can't decide your own expense")
  if (s.status === 'pending_lead' && leadOf(s.userId)?.id !== viewer.id) return forbid('Only the team lead can decide this step')
  if (s.status === 'pending_hr') {
    if (!isHrDecider(viewer)) return forbid('Only HR can decide this step')
    if (s.lead?.byId === viewer.id) {
      return refuse('second_approver_needed', 'You approved this as team lead: someone else in HR gives the second approval')
    }
  }
  if (action === 'reject') {
    if (!reason?.trim()) return invalid('reason', 'string_too_short', 'String should have at least 1 character')
    if (s.status === 'pending_lead') s.lead = { byId: viewer.id, at: now }
    else s.hr = { byId: viewer.id, at: now }
    s.status = 'rejected'
    s.rejection_reason = reason.trim()
  } else if (s.status === 'pending_lead') {
    s.lead = { byId: viewer.id, at: now }
    s.status = 'pending_hr'
  } else {
    s.hr = { byId: viewer.id, at: now }
    s.status = 'approved'
  }
  return { status: 200 as const, expense: toExpense(s) }
}

// Demo data: one expense in every state. Sofia (2) leads João (5) and Marta (6); Helena (18) is HR; Rafael (15) has no lead.
const encode = (text: string) => new TextEncoder().encode(text) as Uint8Array<ArrayBuffer>
const daysAgo = (days: number) => new Date(Date.now() - days * 86_400_000)
const day = (days: number) => daysAgo(days).toISOString().slice(0, 10)
function seed(userId: number, title: string, category: ExpenseCategory, amount: string, ago: number, status: ExpenseStatus, extra: Partial<Stored> = {}) {
  store.push({
    id: nextId++,
    userId,
    title,
    description: null,
    category,
    amount,
    currency: 'EUR',
    spent_on: day(ago + 1),
    status,
    submitted_at: daysAgo(ago).toISOString(),
    lead: null,
    hr: null,
    rejection_reason: null,
    receipts: [{ id: nextReceiptId++, filename: 'receipt.pdf', content_type: 'application/pdf', size: 24, data: encode('%PDF-1.7\nmock receipt\n') }],
    ...extra,
  })
}
seed(5, "Taxi to DKB's office", 'travel', '18.40', 1, 'pending_lead', { description: 'Client workshop, the metro was closed.' })
seed(6, 'Hotel in Frankfurt', 'accommodation', '245.00', 3, 'pending_hr', { lead: { byId: 2, at: daysAgo(2).toISOString() } })
seed(15, 'Conference ticket', 'training', '399.00', 2, 'pending_hr')
seed(5, 'Lunch with the client', 'meals', '36.50', 12, 'approved', {
  lead: { byId: 2, at: daysAgo(11).toISOString() },
  hr: { byId: 18, at: daysAgo(10).toISOString() },
})
seed(5, 'USB-C dock', 'equipment', '89.99', 20, 'rejected', {
  lead: { byId: 2, at: daysAgo(19).toISOString() },
  hr: { byId: 18, at: daysAgo(18).toISOString() },
  rejection_reason: 'Please order equipment through IT.',
})
